overlord/snapstate/snapstate.go

```diff
@@ -201,6 +206,7 @@ func (i pathInfo) SnapSetupForUpdate(st *state.State, params updateParamsFunc, _
 		Type:               i.Type(),
 		PlugsOnly:          len(i.Slots) == 0,
 		InstanceKey:        i.InstanceKey,
+		MigratedHidden:     opts.UseHidden,
```

Comment:
I don't understand all this setting in snapstate.go of MigratedHidden based on UseHidden, they don't seem to match with the definition of the field, which says it's about whether it was migrated during the change itself

Reply:
The idea here was that there's nothing to migrate, so it's already "done". We can just use the hidden dir for any subsequent operations

Reply:
but UseHidden is only the feature flag, did you mean opts.MigratedToHidden? but even then undo will do the wrong thing it will undo something done in a different change

Reply:
Do you mean just this setting in the path install or all the settings including the other Installs? Because I agree that this one is wrong because it's a path install and those works as updates as well. But the idea with the other installs was that the snap data dir doesn't exist yet, so there's nothing to migrate. Unless I'm mistaken and that's not true?

Reply:
I see,  that was not clear at all :) Anyway having to set that inside SnapSetup each time manually is fairly brittle. I looked a bit more at the code there's a couple of simplications that can be done and maybe they can address differently what to do on install:

* we never pass SnapSetup GetSnapDirOpts, we should omit that argument for it
* all the places where we call getDirMigrationOpts itself except the implementation of GetSnapDirOpts already have a snapst SnapState for the snap, so I think it would make sense to:
   * change the signature of getDirMigrationOpts to (*state.State, *SnapState, *SnapSetup) and have only GetSnapDirOpts use Get to find the SnapState to pass in
   * with that change getDirMigrationOpts  can use !snapst.IsInstalled() to decide whether to return a default value for MigratedToHidden based on UseHidden perhaps? which would introduce part of the behavior you were looking for here

Reply:
The simplifications you mention make sense, I've simplified the opts function. 
Regarding the install, I think you're right, these settings were meant as shortcuts but they don't add much value. They make the code less understandable by adding a small difference between install and update and they spread the migration logic in more places. The install can just go through the same process as the updates, since hiding the dir will just skip if there's nothing to hide. I removed the SnapSetup settings and adjusted the tests accordingly. Thanks